docs/source/users/index.md

```diff
@@ -1028,3 +956,93 @@ data:
 runtime:
     /Users/3coins/Library/Jupyter/runtime
 ```
+
+## Custom model providers
+
+You can define new providers using the LangChain framework API. Custom providers
+inherit from both `jupyter-ai`'s ``BaseProvider`` and `langchain`'s [``LLM``][LLM].
+You can either import a pre-defined model from [LangChain LLM list][langchain_llms],
+or define a [custom LLM][custom_llm].
+In the example below, we define a provider with two models using
+a dummy ``FakeListLLM`` model, which returns responses from the ``responses``
+keyword argument.
+
+```python
+# my_package/my_provider.py
+from jupyter_ai_magics import BaseProvider
+from langchain.llms import FakeListLLM
+
+
+class MyProvider(BaseProvider, FakeListLLM):
+    id = "my_provider"
+    name = "My Provider"
+    model_id_key = "model"
+    models = [
+        "model_a",
+        "model_b"
+    ]
+    def __init__(self, **kwargs):
+        model = kwargs.get("model_id")
+        kwargs["responses"] = (
+            ["This is a response from model 'a'"]
+            if model == "model_a" else
+            ["This is a response from model 'b'"]
+        )
+        super().__init__(**kwargs)
+```
+
+
+If the new provider inherits from [``BaseChatModel``][BaseChatModel], it will be available
+both in the chat UI and with magic commands. Otherwise, users can only use the new provider
+with magic commands.
+
+To make the new provider available, you need to declare it as an [entry point](https://setuptools.pypa.io/en/latest/userguide/entry_point.html):
+
+```toml
+# my_package/pyproject.toml
+[project]
+name = "my_package"
+version = "0.0.1"
+
+[project.entry-points."jupyter_ai.model_providers"]
+my-provider = "my_provider:MyProvider"
+```
+
+To test that the above minimal provider package works, install it with:
+
+```sh
+# from `my_package` directory
+pip install -e .
+```
+
+Then, restart JupyterLab. You should now see an info message in the log that mentions
+your new provider's `id`:
+
+```
+[I 2023-10-29 13:56:16.915 AiExtension] Registered model provider `my_provider`.
+```
+
+[langchain_llms]: https://api.python.langchain.com/en/latest/api_reference.html#module-langchain.llms
+[custom_llm]: https://python.langchain.com/docs/modules/model_io/models/llms/custom_llm
+[LLM]: https://api.python.langchain.com/en/latest/llms/langchain.llms.base.LLM.html#langchain.llms.base.LLM
+[BaseChatModel]: https://api.python.langchain.com/en/latest/chat_models/langchain.chat_models.base.BaseChatModel.html
+
+
+### Customizing prompt templates
+
+To modify the prompt template for a given format, override the ``get_prompt_template`` method:
```

Comment:
This is sort of an incomplete feature at the moment, that only works in magic commands and for a specific format. Most users will expect a provider/model specific prompt template that also applies to Chat UI. I am not sure how helpful this feature is at this point, but if we want to keep this, it might be useful to add a small note highlighting this limitation.

This section of the documentation also has reference to the prompt templates. We should probably remove this section from user documentation, as they are unlikely to be able to update these prompt templates.
https://jupyter-ai.readthedocs.io/en/latest/users/index.html#prompt-templates

Reply:
> This is sort of an incomplete feature at the moment, that only works in magic commands and for a specific format. Most users will expect a provider/model specific prompt template that also applies to Chat UI.

Right, this was highlighted by this issue recently: https://github.com/jupyterlab/jupyter-ai/issues/491

> I am not sure how helpful this feature is at this point, but if we want to keep this, it might be useful to add a small note highlighting this limitation.

I think that adding a note is indeed a good way forward. I believe that documenting that customisation of prompt templates for chat is not possible yet (ideally providing a link to issue tracking it) may save advanced users a lot of time.

> This section of the documentation also has reference to the prompt templates. We should probably remove this section from user documentation, as they are unlikely to be able to update these prompt templates.

That subsection is within `%%ai` magic section so I think that this might be fine?

Reply:
@krassowski 
Even though that text is within `%%ai` magic section, users don't have a way to edit the prompt from the notebook. I feel that this info is better placed in the developer or contributor docs where it can be configured.

Reply:
Agree with @3coins that template customization is out of place in the Users section. Moved [## Prompt templates" ](https://jupyter-ai.readthedocs.io/en/latest/users/index.html#prompt-templates) to Developers page, made it a subsection (##). Also moved ### Customizing prompt templates to Developers page as a subsection of  [## Prompt templates" ](https://jupyter-ai.readthedocs.io/en/latest/users/index.html#prompt-templates). Also added a note that this is only available for Jupyter AI magics / not available in the chat and a link to issue #498 tracking it. 

This seems to make the most sense from new reader POV, let me know what do you think.